storage/redis.js

```diff
@@ -0,0 +1,343 @@
+const { findNotMatching, randomElement } = require('../util')
+
+/**
+ * @typedef StorageRedisOptions
+ * @property {!store} client
+ * @property {?Logger} log
+ * @property {?boolean} [invalidation=false]
+ */
+
+class StorageRedis extends StorageInterface {
+  /**
+   * @param {StorageRedisOptions} options
+   */
+  constructor(options) {
+    // TODO validate options
+    // if (!options.client) {
+    //   throw new Error('Redis client is required')
+    // }
+    // if (options.invalidation && !options.listener) {
+    //   throw new Error('Redis listener is required with invalidation')
+    // }
+    super(options)
+    this.options = options
+    this.invalidation = options.invalidation || false
+  }
+
+  async init() {
+    this.log = this.options.log || nullLogger
+    this.store = this.options.client
+
+    if (!this.invalidation) {
+      return
+    }
+
+    // TODO drop
+    // await this.listen(this.options.listener)
+    // this.options.invalidation.clear
+    this.heuristicClean()
+  }
+
+  /**
+   * @param {string} key
+   * @returns {undefined|*} undefined if key not found
+   */
+  async get(key) {
+    this.log.debug({ msg: 'acd/storage/redis.get', key })
+
+    try {
+      const value = await this.store.get(key)
+      if (!value) {
+        // TODO test
+        // clear references because the key could be expired (or evicted)
+        this.clearReferences(key)
+        return undefined
+      }
+      return JSON.parse(value)
+    } catch (err) {
+      this.log.error({ msg: 'acd/storage/redis.get error', err, key })
+    }
+  }
+
+  /**
+   * set value by key
+   * @param {string} key
+   * @param {*} value
+   * @param {number} ttl - ttl in seconds; zero means key will not be stored
+   * @param {?string[]} references
+   */
+  async set(key, value, ttl, references) {
+    // TODO can keys contains * or so?
+    this.log.debug({ msg: 'acd/storage/redis.set key', key, value, ttl, references })
+
+    ttl = Number(ttl)
+    if (!ttl || ttl < 0) {
+      return
+    }
+
+    try {
+      await this.store.set(key, stringify(value), 'EX', ttl)
+
+      if (!references) {
+        return
+      }
+
+      if (!this.invalidation) {
+        this.log.warn({ msg: 'acd/storage/redis.set, invalidation is disabled, references are useless' })
+        return
+      }
+
+      const writes = []
+
+      // clear old references
+      const currentReferences = await this.store.smembers('k:' + key)
+      this.log.debug({ msg: 'acd/storage/redis.set current references', key, currentReferences })
+      if (currentReferences.length > 1) {
+        currentReferences.sort()
+        references.sort()
+        const referencesToRemove = findNotMatching(references, currentReferences)
+
+        // remove key in current references
+        for (const reference of referencesToRemove) {
+          writes.push(['srem', 'r:' + reference, key])
+        }
+        writes.push(['del', 'k:' + key])
+      }
+
```

Comment:
If there are more than 1024 (configurable) elements in the `currentReferences` array, fork the promise chain and check if all elements in the array exist in the database. I would actually recommend using a Lua script as the network latency might wreck this.

Reply:
I'm not a big fan of db scripts, they are not easy to install and maintain. Can we put that as an option?